### PR TITLE
Bugfix for ConfiguredObjectProvenance and StructuralEquals on ConfigurationData

### PR DESCRIPTION
**Pull request by JackSullivan**
## Bugfix

Provenance instances for implementations that depend on `SkeletalConfiguredObjectProvenance` populate an empty string rather than nothing for null-valued string fields that are annotated with `@Config`. This fixes that issue. The issue was discovered because it was causing `ConfigurationData` instances produced through the provenance system to not match those produced through importing a configurable object into `ConfigurationManager`. There may be instances of incorrect provenances in the wild that this affects, but I suspect the chances are fairly small.

## Enhancement

Added a method to check whether two configured objects are structurally equal to one another by checking whether their types and property values match, dereferencing property values that correspond to the names of other configured objects and recursively equality checking them.

### CODE REVIEW

File: olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java [hunk 1]
```diff
@@ -263,6 +274,151 @@ public int hashCode() {
...
+        }
+    }
+
+    /**
+     * Checks whether two ConfigurationData objects are 'structurally equal'. Two objects are structurally
+     * equal when their classNames are the same and when all of their properties are equal or, if those
+     * properties refer to another ConfigurationData by name, if all of their properties are structurally
+     * equal recursively.
+     *
+     * <p></p>
```
Comment (Craigacp) [inline olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:401]: Can you remove the `</p>` close tag?

File: olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java [hunk 2]
```diff
@@ -263,6 +274,151 @@ public int hashCode() {
...
+     *
+     * {@code aName} should be the name of an element of {@code a} that is to be compared to {@code bName}
+     * in {@code b}. {@code a} and {@code b} should each contain all the ConfigurationData objects
+     * needed to instantiate the objects named by {@code aName} and {@code bName} respectively. Objects not
+     * instantiated by traversing children of {@code aName} and {@code bName} are ignored.
+     *
```
Comment (Craigacp) [inline olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:407]: Might be useful to note that this method is extremely chatty at the info level in the Javadoc when it finds discrepancies. Most other things in OLCUT tend to log at fine unless it's an actual issue, and that might be appropriate here.

File: olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java [hunk 3]
```diff
@@ -263,6 +274,151 @@ public int hashCode() {
...
+                                            return eq;
+                                        });
+                            } else if(aProp instanceof MapProperty && bProp instanceof MapProperty) {
+                                Map<String, SimpleProperty> aMap = ((MapProperty) aProp).getMap();
+                                Map<String, SimpleProperty> bMap = ((MapProperty) bProp).getMap();
+                                return aMap.keySet().equals(bMap.keySet()) &&
+                                        aMap.keySet().stream().allMatch(k -> {
+                                            boolean eq = simplePropertyDerefEquals(a, b, aMap.get(k), bMap.get(k), propName);
+                                            if(!eq) {
+                                                logger.info(String.format("Property key: %s, PropValue Key: %s, a.value: %s, b.value: %s",
```
Comment (Craigacp) [inline olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:376]: Doesn't `simplePropertyDerefEquals` already log this? If so, then this lambda can drop the return statement and it'll be much nicer.

File: olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java [hunk 4]
```diff
@@ -263,6 +274,151 @@ public int hashCode() {
...
+            propNames.addAll(bRoot.getProperties().keySet());
+
+            boolean typesMatch = aRoot.getClassName().equals(bRoot.getClassName());
+            if(!typesMatch) {
+                logger.info(String.format("type mismatch, a.class: %s b.class: %s", aRoot.getClassName(), bRoot.getClassName()));
+            }
+
+            return typesMatch &&
+                    propNames.stream().allMatch(propName -> {
+                        Optional<Property> aPropOpt = Optional.ofNullable(aRoot.getProperties().get(propName));
```
Comment (Craigacp) [inline olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:350]: Can we extract this lambda into a static method? It's hard to figure out what it's capturing and that level of complexity in a lambda body worries me (e.g. it's harder to debug). If possible could you recursively do that so the nested lambdas are also static methods? There are too many return statements for too many functions.

**Review by Craigacp — CHANGES_REQUESTED** [review]
Could you also add a test where the equals is true and one where it's false? You should be able to generate objects which round trip through provenance using the existing provenance test utils and then compare to see if they are equal.

**Comment by JackSullivan** [comment]
I believe I've addressed all of the changes addressed above. In the process I removed the tests that used `Random` fields throughout, since they are both deprecated and incompatible with provenance.

File: olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java [hunk 5]
```diff
@@ -263,6 +274,152 @@ public int hashCode() {
...
+                List<SimpleProperty> aList = ((ListProperty) aProp).getSimpleList();
+                List<SimpleProperty> bList = ((ListProperty) bProp).getSimpleList();
+                return aList.size() == bList.size() &&
+                        IntStream.range(0, aList.size()).allMatch(i -> simplePropertyDerefEquals(a, b, aList.get(i), bList.get(i), propName, Optional.of(", List index: " + i)));
+            } else if(aProp instanceof MapProperty && bProp instanceof MapProperty) {
+                Map<String, SimpleProperty> aMap = ((MapProperty) aProp).getMap();
+                Map<String, SimpleProperty> bMap = ((MapProperty) bProp).getMap();
+                return aMap.keySet().equals(bMap.keySet()) &&
+                        aMap.keySet().stream().allMatch(k -> simplePropertyDerefEquals(a, b, aMap.get(k), bMap.get(k), propName, Optional.of(", PropValue key: " + k)));
+            } else {
```
Comment (JackSullivan) [inline olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:346]: Add a clause to account for mismatched property types

File: olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java [hunk 6]
```diff
@@ -263,6 +279,362 @@ public int hashCode() {
...
+                    }) &&
+                            checkPresenceAllMatch(this.listClassProperties, that.listClassProperties, (as, bs) -> {
+                                boolean eq = Util.bagEquality(as, bs);
+                                if(!eq) {
+                                    logger.fine("bags not equal:\na: " + as.toString() +  "\nb: " + bs.toString());
+                                }
+                                return eq;
+                            });
+
+                if(!listMatch) {
```
Comment (Craigacp) [inline olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:414]: This line doesn't seem to be indented properly.

File: olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java [hunk 7]
```diff
@@ -263,6 +279,362 @@ public int hashCode() {
...
+                                                    .allMatch(k -> aMap.get(k).equals(bMap.get(k))));
+
+                    if(!mapMatch) {
+                        logger.fine("MapProperties don't match: as: " + this.mapProperties + " bs: " + that.mapProperties);
+                    }
+
+                    boolean listMatch = checkPresenceAllMatch(this.listProperties,that.listProperties, (as, bs) -> {
+                       boolean eq = Util.bagEquality(as, bs);
+                        if(!eq) {
+                            logger.fine("bags not equal:\na: " + as.toString() +  "\nb: " + bs.toString());
```
Comment (Craigacp) [inline olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:402]: I think this should be `ListProperties not equal`, or `ListProperties not equal using bag equality`. I see there's more logging further down, but I think being specific here and in the class one will make it a little easier to see.

File: olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java [hunk 8]
```diff
@@ -263,6 +279,362 @@ public int hashCode() {
...
+
+        private String reportString() {
+            return String.format("Property Key: %s%s, with value %s %s %s %s %s", propName, locationContext,
+                    innerValue, (isDerefed ? "(is a reference)" : "(is not a reference)"),
+                    innerDoubleValue.map(d -> "(parsed as double " + d +")").orElse("(not parsed as double)"),
+                    innerDateTime.map(d -> "(parsed as date " + d + ")").orElse("(not parsed as date)"),
+                    innerBool.map(d -> "(parsed as bool " + d + ")").orElse("(not parsed as bool)"));
+        }
+
+        @Override public int hashCode() {
```
Comment (Craigacp) [inline olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:506]: We have override on the line before the method declaration.

File: olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java [hunk 9]
```diff
@@ -263,6 +279,362 @@ public int hashCode() {
...
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Checks whether two ConfigurationData objects are 'structurally equal'. Two objects are structurally
+     * equal when their classNames are the same and when all of their properties are equal or, if those
+     * properties refer to another ConfigurationData by name, if all of their properties are structurally
+     * equal recursively.
```
Comment (Craigacp) [inline olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:602]: Add a note saying it does not compare the serializedForm, importable, exportable, leaseTime or entries attributes of the configuration data. Everything apart from serializedForm is going to go away in the next version anyway, but it's worth noting here.

File: olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java [hunk 10]
```diff
@@ -263,6 +279,362 @@ public int hashCode() {
...
+     * <p>
+     *
+     * {@code aName} should be the name of an element of {@code a} that is to be compared to {@code bName}
+     * in {@code b}. {@code a} and {@code b} should each contain all the ConfigurationData objects
+     * needed to instantiate the objects named by {@code aName} and {@code bName} respectively. Objects not
+     * instantiated by traversing children of {@code aName} and {@code bName} are ignored.
+     *
+     * <p>
+     *
+     * At log-level {@code FINE} this reports the first place where the two instances differ, and the nature of their
```
Comment (Craigacp) [inline olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java:621]: Maybe expand this slightly? It logs all the places where the two instances differ in whatever object they differ in. Possibly `reports the first object where the two instances differ`, but that's not particularly clear either.

File: olcut-core/src/test/java/com/oracle/labs/mlrg/olcut/config/ConfigurationDataTest.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
+package com.oracle.labs.mlrg.olcut.config;
```
Comment (Craigacp) [inline olcut-core/src/test/java/com/oracle/labs/mlrg/olcut/config/ConfigurationDataTest.java:1]: This file needs a copyright & license header.

File: olcut-core/src/test/java/com/oracle/labs/mlrg/olcut/config/ConfigurationDataTest.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        Logger logger = Logger.getLogger(ConfigurationData.class.getName());
+        logger.setLevel(Level.FINE);
+    }
+
+
+    @Test
+    public void structuralEqualsAllConfig() {
+
+        ConsoleHandler ch = new ConsoleHandler();
+        ch.setLevel(Level.FINE);
```
Comment (Craigacp) [inline olcut-core/src/test/java/com/oracle/labs/mlrg/olcut/config/ConfigurationDataTest.java:31]: Can you back out the logging changes from this test?

**Review by Craigacp — CHANGES_REQUESTED** [review]
A bunch of small things, mostly formatting and docs.

**Review by Craigacp — APPROVED** [review]
LGTM

**Pull request merged**